pyartcd/pyartcd/locks.py

```diff
@@ -143,6 +150,34 @@ async def get_locks(self, version: str = None):
         return await redis.get_keys(pattern)
 
 
+async def enqueue_for_lock(coro: coroutine, lock: Lock, lock_name: str, lock_id: str,
+                           ocp_version: str, version_queue_name):
+    lock_manager = LockManager.from_lock(lock)
+    return await _enqueue_for_lock(coro, lock_manager, lock, lock_name, lock_id, ocp_version, version_queue_name)
+
+
+@tenacity.retry(
+    wait=tenacity.wait_fixed(600),  # wait for 10 minutes between retries
+    stop=tenacity.stop_after_attempt(600 * 24),  # wait for 24 hours
+    retry=tenacity.retry_if_exception_type(),
+)
+async def _enqueue_for_lock(coro: coroutine, lock_manager, lock: Lock, lock_name: str, lock_id: str,
+                            ocp_version: str, version_queue_name):
+    if not await lock_manager.is_locked(lock_name):
+        # TODO: use a redis tx here
+        # fetch the first element in the reverse sorted set by score (item with the max score)
+        result = await redis.call('zrange', version_queue_name, 0, 0, desc=True)
+        # there should be at least 1 entry, if not error out
+        version_on_top = result[0]
+        if version_on_top == ocp_version:
```

Comment:
this only contains ocp_version, what will happen if trigger two 4.13 jobs will it follow the job's order?

Reply:
Yeah, this was something of a design decision. Since this is a sorted set, the item that we insert is `{string: int/float}`. We can make the string `ocp_version#jobid` or similar and make sure the job that got in the queue is the job that gets popped. But that introduces us to situations where when the job reaches timeout (which will frequently happen since these are very long running jobs), we will need to pop it out of the queue in some way. But then we also lose our place in the queue. If between the interval of one ocp4 4.14 job reaching timeout popping itself, and then next instance of it inserting itself, the mass rebuild lock can be acquired by a 4.13 job still in the queue.

Rather than that, having generic not job specific ids, let us keep our place in queue.